File: web/src/api/myPortfolio.ts
/**
 * My Portfolio Report API client
 */
import client from './client';

export interface PortfolioSummary {
    total_models: number;
    action_items_count: number;
    overdue_count: number;
    compliant_percentage: number;
    models_compliant: number;
    models_non_compliant: number;
    yellow_alerts: number;
    red_alerts: number;
}

export interface ActionItem {
    type: 'attestation' | 'recommendation' | 'validation_submission';
    urgency: 'overdue' | 'in_grace_period' | 'due_soon' | 'upcoming' | 'unknown';
    model_id: number;
    model_name: string;
    item_id: number;
    item_code: string | null;
    title: string;
    action_description: string;
    due_date: string | null;
    days_until_due: number | null;
    link: string;
}

export interface MonitoringAlert {
    model_id: number;
    model_name: string;
    metric_name: string;
    metric_value: number | null;
    qualitative_outcome: string | null;
    outcome: 'YELLOW' | 'RED';
    cycle_name: string;
    cycle_id: number;
    plan_id: number;
    result_date: string;
}

export interface CalendarItem {
    due_date: string;
    type: 'attestation' | 'recommendation' | 'validation_submission';
    model_id: number;
    model_name: string;
    item_id: number;
    item_code: string | null;
    title: string;
    is_overdue: boolean;
}

export interface PortfolioModel {
    model_id: number;
    model_name: string;
    risk_tier: string | null;
    risk_tier_code: string | null;
    approval_status: string | null;
    approval_status_code: string | null;
    last_validation_date: string | null;
    next_validation_due: string | null;
    days_until_due: number | null;
    open_recommendations: number;
    attestation_status: string | null;
    yellow_alerts: number;
    red_alerts: number;
    has_overdue_items: boolean;
    ownership_type: 'primary' | 'shared' | 'delegate';
}

export interface MyPortfolioResponse {
    report_generated_at: string;
    as_of_date: string;
    team_id: number | null;
    team_name: string;
    summary: PortfolioSummary;
    action_items: ActionItem[];
    monitoring_alerts: MonitoringAlert[];
    calendar_items: CalendarItem[];
    models: PortfolioModel[];
}

export async function getMyPortfolio(teamId?: number): Promise<MyPortfolioResponse> {
    const params = new URLSearchParams();
    if (teamId !== undefined) {
        params.append('team_id', String(teamId));
    }
    const suffix = params.toString() ? `?${params.toString()}` : '';
    const response = await client.get<MyPortfolioResponse>(`/reports/my-portfolio${suffix}`);
    return response.data;
}

/**
 * Export portfolio to PDF via backend API
 */
export async function exportPortfolioToPDF(): Promise<Blob> {
    const response = await client.get('/reports/my-portfolio/pdf', {
        responseType: 'blob',
    });
    return response.data;
}

/**
 * Get urgency badge color class based on urgency level
 */
export function getUrgencyColorClass(urgency: string): string {
    switch (urgency) {
        case 'overdue':
            return 'bg-red-100 text-red-800';
        case 'in_grace_period':
            return 'bg-orange-100 text-orange-800';
        case 'due_soon':
            return 'bg-yellow-100 text-yellow-800';
        case 'upcoming':
            return 'bg-blue-100 text-blue-800';
        default:
            return 'bg-gray-100 text-gray-800';
    }
}

/**
 * Get urgency label for display
 */
export function getUrgencyLabel(urgency: string): string {
    switch (urgency) {
        case 'overdue':
            return 'Overdue';
        case 'in_grace_period':
            return 'Grace Period';
        case 'due_soon':
            return 'Due Soon';
        case 'upcoming':
            return 'Upcoming';
        default:
            return 'Unknown';
    }
}

/**
 * Get action type icon
 */
export function getActionTypeIcon(type: string): string {
    switch (type) {
        case 'attestation':
            return '⏰';
        case 'recommendation':
            return '📋';
        case 'validation_submission':
            return '📝';
        default:
            return '📌';
    }
}

/**
 * Get ownership type label
 */
export function getOwnershipLabel(type: string): string {
    switch (type) {
        case 'primary':
            return 'Owner';
        case 'shared':
            return 'Shared Owner';
        case 'delegate':
            return 'Delegate';
        default:
            return type;
    }
}

/**
 * Format days until due for display
 */
export function formatDaysUntilDue(days: number | null): string {
    if (days === null) return 'No date';
    if (days === 0) return 'Today';
    if (days === 1) return 'Tomorrow';
    if (days === -1) return '1 day ago';
    if (days < 0) return `${Math.abs(days)} days ago`;
    return `In ${days} days`;
}

/**
 * Export portfolio to CSV
 */
export function exportPortfolioToCSV(report: MyPortfolioResponse): string {
    const lines: string[] = [];

    // Header
    lines.push('My Model Portfolio Report');
    lines.push(`Generated: ${report.report_generated_at}`);
    lines.push(`As of: ${report.as_of_date}`);
    lines.push('');

    // Summary
    lines.push('SUMMARY');
    lines.push(`Total Models,${report.summary.total_models}`);
    lines.push(`Action Items,${report.summary.action_items_count}`);
    lines.push(`Overdue Items,${report.summary.overdue_count}`);
    lines.push(`Compliance Rate,${report.summary.compliant_percentage}%`);
    lines.push(`Yellow Alerts,${report.summary.yellow_alerts}`);
    lines.push(`Red Alerts,${report.summary.red_alerts}`);
    lines.push('');

    // Action Items
    lines.push('ACTION ITEMS');
    lines.push('Type,Urgency,Model,Item Code,Title,Action,Due Date,Days Until Due');
    for (const item of report.action_items) {
        lines.push([
            item.type,
            item.urgency,
            `"${item.model_name}"`,
            item.item_code || '',
            `"${item.title}"`,
            `"${item.action_description}"`,
            item.due_date || '',
            item.days_until_due?.toString() || '',
        ].join(','));
    }
    lines.push('');

    // Monitoring Alerts
    lines.push('MONITORING ALERTS');
    lines.push('Model,Metric,Value,Outcome,Cycle,Date');
    for (const alert of report.monitoring_alerts) {
        lines.push([
            `"${alert.model_name}"`,
            `"${alert.metric_name}"`,
            alert.metric_value?.toString() || alert.qualitative_outcome || '',
            alert.outcome,
            alert.cycle_name,
            alert.result_date,
        ].join(','));
    }
    lines.push('');

    // Models
    lines.push('MODEL PORTFOLIO');
    lines.push('Model ID,Model Name,Risk Tier,Approval Status,Last Validation,Open Recs,Yellow Alerts,Red Alerts,Ownership');
    for (const model of report.models) {
        lines.push([
            model.model_id.toString(),
            `"${model.model_name}"`,
            model.risk_tier || '',
            model.approval_status || '',
            model.last_validation_date || '',
            model.open_recommendations.toString(),
            model.yellow_alerts.toString(),
            model.red_alerts.toString(),
            model.ownership_type,
        ].join(','));
    }

    return lines.join('\n');
}

File: api/app/schemas/my_portfolio.py
"""My Portfolio Report schemas for model owner dashboard."""
from datetime import datetime, date
from typing import Optional, List
from pydantic import BaseModel, Field


class PortfolioSummary(BaseModel):
    """Summary statistics for the portfolio."""
    total_models: int = Field(...,
                              description="Total number of models in portfolio")
    action_items_count: int = Field(...,
                                    description="Number of items requiring action")
    overdue_count: int = Field(..., description="Number of overdue items")
    compliant_percentage: float = Field(
        ..., description="Percentage of models that are compliant")
    models_compliant: int = Field(...,
                                  description="Number of compliant models")
    models_non_compliant: int = Field(...,
                                      description="Number of non-compliant models")
    yellow_alerts: int = Field(...,
                               description="Number of yellow monitoring alerts")
    red_alerts: int = Field(..., description="Number of red monitoring alerts")
    open_exceptions_count: int = Field(
        0, description="Number of open model exceptions")


class ActionItem(BaseModel):
    """An item requiring user action."""
    type: str = Field(...,
                      description="Type: attestation, recommendation, validation_submission")
    urgency: str = Field(...,
                         description="Urgency: overdue, in_grace_period, due_soon, upcoming")
    model_id: int = Field(..., description="ID of the related model")
    model_name: str = Field(..., description="Name of the related model")
    item_id: int = Field(..., description="ID of the specific item")
    item_code: Optional[str] = Field(
        None, description="Code/reference (e.g., REC-2025-00042)")
    title: str = Field(..., description="Brief title of the action needed")
    action_description: str = Field(...,
                                    description="Description of the action required")
    due_date: Optional[date] = Field(
        None, description="Due date for the action")
    days_until_due: Optional[int] = Field(
        None, description="Days until due (negative if overdue)")
    link: str = Field(..., description="Navigation path to the item")


class MonitoringAlert(BaseModel):
    """A monitoring result with yellow or red outcome."""
    model_id: int = Field(..., description="ID of the model")
    model_name: str = Field(..., description="Name of the model")
    metric_name: str = Field(..., description="Name of the KPM metric")
    metric_value: Optional[float] = Field(
        None, description="Numeric value (if quantitative)")
    qualitative_outcome: Optional[str] = Field(
        None, description="Qualitative outcome label")
    outcome: str = Field(..., description="Calculated outcome: YELLOW or RED")
    cycle_name: str = Field(..., description="Name of the monitoring cycle")
    cycle_id: int = Field(..., description="ID of the monitoring cycle")
    plan_id: int = Field(..., description="ID of the monitoring plan")
    result_date: date = Field(..., description="Date the result was recorded")


class ExceptionItem(BaseModel):
    """An open model exception requiring attention."""
    exception_id: int = Field(..., description="Exception ID")
    exception_code: str = Field(...,
                                description="Exception code (e.g., EXC-2025-00001)")
    exception_type: str = Field(
        ..., description="Type: UNMITIGATED_PERFORMANCE, OUTSIDE_INTENDED_PURPOSE, USE_PRIOR_TO_VALIDATION")
    exception_type_label: str = Field(...,
                                      description="Human-readable exception type")
    model_id: int = Field(..., description="ID of the model")
    model_name: str = Field(..., description="Name of the model")
    status: str = Field(..., description="Status: OPEN or ACKNOWLEDGED")
    description: str = Field(..., description="Exception description")
    detected_at: datetime = Field(...,
                                  description="When the exception was detected")
    acknowledged_at: Optional[datetime] = Field(
        None, description="When acknowledged (if applicable)")
    link: str = Field(..., description="Navigation path to the model")


class CalendarItem(BaseModel):
    """A deadline item for calendar view."""
    due_date: date = Field(..., description="Due date")
    type: str = Field(...,
                      description="Type: attestation, recommendation, validation_submission")
    model_id: int = Field(..., description="ID of the related model")
    model_name: str = Field(..., description="Name of the related model")
    item_id: int = Field(..., description="ID of the specific item")
    item_code: Optional[str] = Field(
        None, description="Code/reference if applicable")
    title: str = Field(..., description="Brief title")
    is_overdue: bool = Field(..., description="Whether the item is past due")


class PortfolioModel(BaseModel):
    """A model in the user's portfolio."""
    model_id: int = Field(..., description="Model ID")
    model_name: str = Field(..., description="Model name")
    risk_tier: Optional[str] = Field(
        None, description="Risk tier label (e.g., 'Tier 1')")
    risk_tier_code: Optional[str] = Field(None, description="Risk tier code")
    approval_status: Optional[str] = Field(
        None, description="Model approval status label (e.g., 'Approved', 'Expired')")
    approval_status_code: Optional[str] = Field(
        None, description="Model approval status code (e.g., 'APPROVED', 'EXPIRED')")
    last_validation_date: Optional[date] = Field(
        None, description="Date of last validation")
    next_submission_due: Optional[date] = Field(
        None, description="Next submission due date")
    next_validation_due: Optional[date] = Field(
        None, description="Next validation due date")
    days_until_submission_due: Optional[int] = Field(
        None, description="Days until next submission due")
    days_until_validation_due: Optional[int] = Field(
        None, description="Days until next validation due")
    open_recommendations: int = Field(
        0, description="Count of open recommendations")
    attestation_status: Optional[str] = Field(
        None, description="Current attestation status")
    yellow_alerts: int = Field(
        0, description="Count of yellow monitoring alerts")
    red_alerts: int = Field(0, description="Count of red monitoring alerts")
    open_exceptions: int = Field(
        0, description="Count of open model exceptions")
    has_overdue_items: bool = Field(
        False, description="Whether model has any overdue items")
    ownership_type: str = Field(
        ..., description="How user owns: primary, shared, developer, delegate")


class MyPortfolioResponse(BaseModel):
    """Complete portfolio report response."""
    report_generated_at: datetime = Field(...,
                                          description="Timestamp when report was generated")
    as_of_date: date = Field(...,
                             description="Date the metrics are calculated as of")
    team_id: Optional[int] = Field(
        None, description="Team ID if filtered, 0 for Unassigned, None for all")
    team_name: str = Field(
        "All Teams", description="Team name or 'All Teams' if no filter")

    summary: PortfolioSummary = Field(..., description="Summary statistics")
    action_items: List[ActionItem] = Field(
        ..., description="Items requiring action, sorted by urgency")
    monitoring_alerts: List[MonitoringAlert] = Field(
        ..., description="Yellow/red monitoring results")
    open_exceptions: List[ExceptionItem] = Field(
        ..., description="Open model exceptions requiring attention")
    calendar_items: List[CalendarItem] = Field(
        ..., description="All deadlines for calendar view")
    models: List[PortfolioModel] = Field(...,
                                         description="All models in portfolio")

    model_config = {"from_attributes": True}
